WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,105 @@
 using Microsoft.Extensions.Hosting;
 using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
 using WalletWasabi.Logging;
 
 namespace WalletWasabi.Bases
 {
+	/// <summary>
+	/// <see cref="PeriodicRunner"/> is an extension of <see cref="BackgroundService"/> that is useful for tasks
+	/// that are supposed to repeat regularly.
+	/// </summary>
 	public abstract class PeriodicRunner : BackgroundService
 	{
 		protected PeriodicRunner(TimeSpan period)
 		{
-			TriggeringCts = new CancellationTokenSource();
-			TriggerLock = new object();
 			Period = period;
-			ResetLastException();
+			ExceptionTracker = new LastExceptionTracker();
 		}
 
-		private CancellationTokenSource TriggeringCts { get; set; }
-		private object TriggerLock { get; }
 		public TimeSpan Period { get; }
-		public Exception LastException { get; set; }
-		public long LastExceptionCount { get; set; }
-		public DateTimeOffset LastExceptionFirstAppeared { get; set; }
 
-		private void ResetLastException()
-		{
-			LastException = null;
-			LastExceptionCount = 0;
-			LastExceptionFirstAppeared = DateTimeOffset.MinValue;
-		}
+		private volatile TaskCompletionSource<bool>? _tcs;
```

Comment:
This field should be placed before the constructor.

Reply:
Why does CodeFactor not report that?

Reply:
I have no idea.